CHANGELOG.md

```diff
@@ -1,5 +1,27 @@
+  * The `json` and `cbor` multicodec names (as used by `input-enc` and `format` options) are now no longer aliases for `dag-json` and `dag-cbor` respectively. Instead, they now refer to their proper [multicodec](https://github.com/multiformats/multicodec/blob/master/table.csv) types. `cbor` refers to a plain CBOR format, which will not encode CIDs and does not have strict deterministic encoding rules. `json` is a plain JSON format, which also won't encode CIDs and will encode bytes in the Go-specific padded base64 string format rather than the dag-json method of byte encoding. See https://ipld.io/specs/codecs/ for more information on IPLD codecs.
+* The **dag-pb codec**, which is used to encode UnixFS data for IPFS, is now represented in a form via the `dag` API that mirrors the protobuf schema used to define the binary format and unifies the implementations and specification of dag-pb across the IPLD and IPFS stacks. Previously, additional layers of code within IPFS between protobuf serialization and UnixFS handling for file and directory data, obscured the forms that are described by the protobuf representation. Much of this code has now been replaced and there are fewer layers of transformation. This means that interacting with dag-pb data via the `dag` API will use different forms:
+  * Previously, using `dag get` on a dag-pb block would present the block serialized as JSON as `{"data":"padded-base64-bytes","links":[{"Name":"foo","Size":100,"Cid":{"/":"Qm..."}},...]}`.
+  * Using the dag-pb data model specification and the new default dag-json codec for output, this will now be serialized as: `{"Data":{"/":{"bytes":"unpadded-base64-bytes"}},"Links":[{"Name":"foo","Tsize":100,"Hash":{"/":"Qm..."}},...]}`. Aside from the change in byte formatting, most field names have changed: `data` → `Data`, `links` → `Links`, `Size` → `Tsize`, `Cid` → `Hash`. Note that this output can be changed now using the `--format` option to specify an alternative codec.
+  * Using `dag put` and a `format` option of `dag-pb` now requires that the input conform to this dag-pb specified form. Previously, input using `{"data":"...","links":[...]}` was accepted, now it must be `{"Data":"...","Links":[...]}`.
+  * Previously it was not possible to use paths to navigate to any of these properties of a dag-pb node, the only possible paths were named links, e.g. `dag get QmFoo/NamedLink` where `NamedLink` was one of the links whose name was `NamedLink`. This functionality remains the same, but by prefixing the path with `/ipld/` we enter data model pathing semantics and can `dag get /ipld/QmFoo/Links/0/Hash` to navigate to links or `/ipld/QmFoo/Data` to simply retrieve the data section of the node, for example.
```

Comment:
> e.g. `dag get QmFoo/NamedLink` where `NamedLink` was one of the links whose name was `NamedLink`. This functionality remains the same

This is true in many cases, but IMO is not something we want to guarantee. If we happen to make the UnixFS reifier stricter such that `/ipfs/QmNonUnixFSDagPbObject/LinkName` no longer resolved (and you needed to go the `/Links/0` route) that doesn't seem to scary to me (i.e. we don't need to write lots of path resolution tests for non-UnixFS DagPB objects to verify `/ipfs/QmNonUnixFSDagPbObject/LinkName` still works).

@rvagg WDYT, is it safer to not mention the fate `/ipfs/QmNonUnixFSDagPbObject/LinkName` here?

Reply:
> `/ipfs/QmNonUnixFSDagPbObject/LinkName`

☝️ I believe this is literally what @gozala's niftysave meta-wrapper does, that is: you have users in he wild.

/cc @mikeal ( I think Gozala is OOO )

Reply:
@ribasushi IIRC that uses DAG-CBOR not DagPB so is both not Qm and also not a problem here.

Reply:
Ah I misread the `NonUnixFSDagPb` part.
You are correct, my comment does not apply

Reply:
@aschmahmann I think this is your call since you have a much clearer picture of the direction of IPFS. We have a long-standing note about this ["alternative pathing"](https://ipld.io/specs/codecs/dag-pb/spec/#alternative-legacy-pathing) in the dag-pb spec and I've always just considered it a legacy hold-over, like so many of the dag-pb decisions, that we're living with. I don't find `QmFoo/NamedLink` particularly offensive as a legacy convenience holdover but it really fits in the ADL category in this new world. Having `PathedPBNode` sit in between seem like a reasonable compromise and even if `UnixFSPBNode` evolves to be much smarter and stricter, go-unixfsnode could still retain this fallaback to `PathedPBNode` that will _just_ do named pathing. But you're the one with a better imagination for what future evolution might involve here. I've just seen notes throughout go-unixfsnode alluding to that suggestion and I'm not sure what that really means.

This is the bit that's doing the fallback now, perhaps it just stays and even if `DecodeUnixFSData()` starts rejecting more things we still get the fallback to `PathedPBNode` (which is the `defaultReifier()`)? https://github.com/ipfs/go-unixfsnode/blob/f6007e84b8c4d0d910a11c5380c3e06eeef06403/reification.go#L25-L29

Your call.

Reply:
Thanks @rvagg. It's a tough call, but I think for now we'll leave it as you suggested (i.e. just telling people about the existence of data model pathing and not explicitly dissuading the use of `PathedPBNode` pathing). 

Hopefully there aren't too many people using non-UnixFS DagPB right now and relying on this functionality, but we can push the discussion of if there are any downsides to leaving `PathedPBNode` pathing as a fallback for `/ipfs` for another day. In the meanwhile we're not breaking anyone so there's not much to say here.